File: conquer3d/csrc/binds/creation/triangle_creation.cpp
#include <torch/extension.h>
#include <pybind11/pybind11.h>
#include "../../creation/triangle_creation.h"

namespace py = pybind11;

void bind_creation_triangle_creation(py::module_ &m) {
    m.def("create_sphere", &triangle_creation::create_sphere,
          py::arg("sectors") = 32, py::arg("stacks") = 16, py::arg("radius") = 1.0f,
          R"pbdoc(
          Generates a parameterized UV sphere mesh on CPU.

          Args:
              sectors (int, optional): Number of longitudinal angular slices. Defaults to 32.
              stacks (int, optional): Number of latitudinal vertical stacks. Defaults to 16.
              radius (float, optional): Radius of the sphere. Defaults to 1.0.

          Returns:
              Tuple[torch.Tensor, torch.Tensor]:
                  - vertices (torch.Tensor): ((sectors-1)*stacks + 2, 3) float32 coordinates on CPU.
                  - triangles (torch.Tensor): (2*(stacks-1)*sectors, 3) int32 triangle vertex indices on CPU.

          Example:
              >>> import torch
              >>> from conquer3d._C import create_sphere
              >>> verts, tris = create_sphere(sectors=32, stacks=16, radius=1.0)
          )pbdoc");
    m.def("create_tetrahedra", &triangle_creation::create_tetrahedra,
          py::arg("radius") = 1.0f,
          R"pbdoc(
          Generates a regular 4-faced tetrahedron mesh inscribed in a sphere of given radius on CPU.

          Args:
              radius (float, optional): Circumscribed sphere radius. Defaults to 1.0.

          Returns:
              Tuple[torch.Tensor, torch.Tensor]:
                  - vertices (torch.Tensor): (4, 3) float32 vertex coordinates on CPU.
                  - triangles (torch.Tensor): (4, 3) int32 triangle vertex indices on CPU.

          Example:
              >>> import torch
              >>> from conquer3d._C import create_tetrahedra
              >>> verts, tris = create_tetrahedra(radius=1.0)
          )pbdoc");
}

File: conquer3d/creation/triangle_creation.py
"""Procedural 3D canonical primitive generation routines.

This module provides parametric generators for standard canonical primitives
such as UV spheres and regular tetrahedra.
"""

from typing import Tuple
import torch

from .._C import create_sphere as create_sphere_c
from .._C import create_tetrahedra as create_tetrahedra_c


def create_sphere(
    sectors: int = 32,
    stacks: int = 16,
    radius: float = 1.0
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Generates a parametric UV sphere mesh with vertex coordinates and triangle faces.

    Args:
        sectors (int, optional): Number of longitudinal angular sectors. Defaults to 32.
        stacks (int, optional): Number of latitudinal horizontal stacks. Defaults to 16.
        radius (float, optional): Radius of the generated sphere. Defaults to 1.0.

    Returns:
        Tuple[torch.Tensor, torch.Tensor]:
            - vertices (torch.Tensor): Float32 tensor of shape `(N, 3)` with 3D vertex positions.
            - triangles (torch.Tensor): Int32 tensor of shape `(M, 3)` with triangle corner indices.

    Example:
        >>> from conquer3d.creation import create_sphere
        >>> verts, faces = create_sphere(sectors=64, stacks=32, radius=1.0)
    """
    return create_sphere_c(sectors, stacks, radius)


def create_tetrahedra(
    radius: float = 1.0
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Generates a regular canonical tetrahedron mesh.

    Args:
        radius (float, optional): Circumscribed radius of the tetrahedron. Defaults to 1.0.

    Returns:
        Tuple[torch.Tensor, torch.Tensor]:
            - vertices (torch.Tensor): Float32 tensor of shape `(4, 3)` with 4 vertex coordinates.
            - triangles (torch.Tensor): Int32 tensor of shape `(4, 3)` with 4 outward triangle faces.

    Example:
        >>> from conquer3d.creation import create_tetrahedra
        >>> verts, faces = create_tetrahedra(radius=1.0)
    """
    return create_tetrahedra_c(radius)
